enhancements/insights/pulling-data-from-ocm.md

```diff
@@ -0,0 +1,96 @@
+---
+title: pulling-and-exposing-data-from-ocm
+authors:
+  - "@tremes"
+reviewers:
+  - "@sbose78"
+  - "@0sewa0"
+  - "@inecas"
+  - "@smarterclayton"
+approvers:
+  - "@sbose78"
+  - "@smarterclayton"
+creation-date: 2021-03-04
+last-updated: 2021-03-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Insights Operator pulling and exposing data from the OCM API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will enable the Insights Operator to pull the data (SCA certs)
+from the OCM (OpenShift Cluster Manager) API. The data will be exposed by the Insights Operator
+in the OpenShift API to allow users to use them when consuming and building container images
+on the platform.
+
+## Motivation
+
+Users could consume RHEL content and container images using the RHEL subscription in the OpenShift 3.x.
+In the OpenShift 4, this is no longer possible because the Red Hat Enterprise Linux Core OS (RHCOS) does not
+provide any attached subscription. This enhancement is to provide users the Simple Content Access (SCA) certs
+from Red Hat Subscription Manager (RHSM).
+
+### Goals
+
+- Extend the Insights Operator config with an OCM API URL to be able to query the data
```

Comment:
Why is it the responsibility of Insights Operator to do this?

Reply:
I can't answer that. I was not present during the original decision. So perhaps @iNecas or @smarterclayton.

Reply:
@smarterclayton FYI :-) I would be interested as well to know why was this embedded into insights operator

Reply:
FYI, this effort has been made after consulting with and on direct recommendation from @smarterclayton: If I understood it correctly, the idea is for the insights operator to cover more the "value of RH subscription": where the entitlement certificates fit as well and it seemed there was no better place to put it in. Also, insights-operator already has a trusted connection established with cloud.redhat.com api, where the data are being pulled from.

Reply:
Also @dhellmann, @mfojtik  feel free talking to Siamak Sadeghianfar, that is the PM owning this feature AFAIK

Reply:
@iNecas well, I think openshift controller manager operator might be a better fit for this as this feature will be mainly be used by builds, right? also the way this gets injected into build containers will probably become something in build API?

/cc @adambkaplan

Reply:
@sbose78 is driving this functionality from the building perspective AFAIK (pulling him in).

Reply:
> I think openshift controller manager operator might be a better fit for this as this feature will be mainly be used by builds, right? 

I had the very same question to @smarterclayton when this idea first came and he felt quite convinced IO was the right place compared to all alternatives.

Reply:
SCA Certs (and entitlement certs in general) are a Red Hat subscription experience. Insights operator is the only core OCP component we have today that connects OpenShift to the rest of the Red Hat experience. Correct me if I'm wrong, but I don't think the insights operator provides any value/function without a Red Hat subscription. Openshift-controller-manager, on the other hand, does not need a Red Hat subscription to function.

Builds are a consumer of the SCA cert, but it is by no means the only workload that could find itself needing access to subscription content. There are other components in the works which are designed to mount the cluster's SCA cert into any workload, namely the shared resource/projected resource CSI driver.

Reply:
> also the way this gets injected into build containers will probably become something in build API?

Correct, however through mechanisms that are more general purpose:

1. The projected resource CSI driver will allow the SCA cert to be accessed in the cluster through the `Share` custom resource.
2. The build API will be enhanced to allow `CSI` volumes to be mounted in.

Reply:
@adambkaplan @iNecas it would be great if the body of this enhancement included the explanation you've given in this thread.

Reply:
Build API changes to support volume mounts
https://github.com/openshift/api/pull/954/files

Reply:
> Builds are a consumer of the SCA cert, but it is by no means the only workload that could find itself needing access to subscription content.

+1 

Subscription content is something that any workload in Openshift should be able to use. Builds happen to be a specialized case.

Reply:
>  Insights operator is the only core OCP component we have today that connects OpenShift to the rest of the Red Hat experience.

Correct.

Additionally, the general intent is to streamline all OCM API communications through this component. In that sense, the term "Insights" isn't quite indicative of what the future of Insights Operator would look like :)

Reply:
OK thanks everybody. Let me try to sum the main reasons there.

Reply:
I think the "insights" operator is kind of confusing name then. I thought the purpose of "insights" is to provide diagnostic data to RH so we can better see cluster issues and we can detect classes of problems and provide mitigations.

Now the insights operator will have second role which is pulling a cert material from RH that enable subscriptions to be used in builds. I'm little concerned about mixing the purposes here.

Reply:
> @adambkaplan @iNecas it would be great if the body of this enhancement included the explanation you've given in this thread.

I don't see the info from this thread in the doc, yet.

Reply:
> Now the insights operator will have second role which is pulling a cert material from RH that enable subscriptions to be used in builds. I'm little concerned about mixing the purposes here.

I agree it's slightly awkward, but since it's the same team either way, I don't care all that much how they manage the deployment of the stuff they are responsible for (one component with 2 jobs, or 2 components with 2 jobs).  It may make it slightly less obvious for someone to find the right component to debug if things aren't working right, but i dunno that that cost justifies the exercise of defining a whole new operator component to do this one trivial operation.

So as long as the reasoning is included in the EP, i think we can push past this discussion.  If this were one team trying to push code into another team's component(when it clearly didn't belong), i'd feel differently.

Reply:
It would be good to have a few sentences in the alternatives section explaining that a new operator could have been created but it would require more work from multiple teams (building new images, etc.) and would require more CPU and RAM resources in the cluster. The counter argument is that it makes the cluster operator status more focused, but I think that's not worth the extra work, myself.

Reply:
OK why not, but I think Michal was suggesting to move this functionality to `openshift-controller-manager` (This is briefly mentioned there). I can also mention this third option of completely new component/operator.